File: backend/app/schemas/transfers.py
"""Wire models for the manual transfer picker (ADR-0008/0018).

Linking already has its shapes in ``schemas.transactions`` (``TransferLink`` /
``TransferOut``), frozen with the transaction contract. What lives here is the
*read* the picker needs before a link exists: which rows could be the other leg,
and what pairing each one would cost.
"""

from __future__ import annotations

import uuid
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.transactions import TransactionOut


class TransferCandidateOut(BaseModel):
    """One possible counterpart leg, with the price of pairing it.

    ``fx_cost_base`` is what the transfer group's ``fx_cost_base`` would become if
    the user committed — Σ base_amount of the two legs. So it is ``None`` for a
    clean same-currency pair (the legs cancel exactly) and the real spread for a
    cross-currency one (ADR-0018). It is computed by the same helper that
    ``link_transfer`` stores, so the number shown before linking is the number
    that lands on the group after it.

    ``within_tolerance`` says the pair's base amounts cancel inside the matching
    tolerance. False is not a refusal: ADR-0018 keeps explicit user linking as the
    override, so the row is still offered — this flag plus the residual are how
    the UI tells "the bank's rate differed slightly" apart from "that is probably
    not the other leg at all".
    """

    transaction: TransactionOut
    # Whole days between the legs, floored — the window is symmetric, so this is
    # never negative.
    days_apart: int
    fx_cost_base: Decimal | None
    within_tolerance: bool


class TransferCandidatesOut(BaseModel):
    items: list[TransferCandidateOut]


class TransferDetailOut(BaseModel):
    """A linked transfer read back whole — the group, and both legs with it.

    A transaction knows only its ``transfer_group_id``, so this is the only way a
    detail view can show the *other* leg. ``fx_cost_base`` is part of that read
    rather than something the client derives, because ADR-0018 requires the cost
    of a cross-currency transfer to be shown: leaving it out would make hiding it
    the path of least resistance.
    """

    transfer_group_id: uuid.UUID
    matched_by: str
    fx_cost_base: Decimal | None
    # Ordered by date then id — deterministic, with no claim about which leg is
    # "the" transfer. A caller picks the side it wants by the amount's sign.
    legs: list[TransactionOut]

File: frontend/src/api/transfers.ts
// The transfer vertical's own API surface: propose a counterpart leg, link it,
// unlink it, and read a linked pair back.
//
// Kept out of api/hooks.ts (shared, and pinned by other workstreams) while
// following its conventions — including reusing the shared invalidation list, so
// a link or an unlink refreshes the reports exactly like any other ledger write.
// That matters more here than elsewhere: linking changes what cash-flow and
// spending contain, not just how one row looks.

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "@/api/client";
import { useInvalidateLedger } from "@/api/hooks";
import type { Money, Transaction, UUID } from "@/api/types";

/** One row that could be the other leg, priced before the user commits. */
export interface TransferCandidate {
  transaction: Transaction;
  /** Whole days between the legs; the window is symmetric, so never negative. */
  days_apart: number;
  /**
   * What the pair's `fx_cost_base` would become: the real FX spread for a
   * cross-currency transfer, null when the legs cancel exactly (or when a rate
   * is missing). Shown *before* linking — the whole point of ADR-0018.
   */
  fx_cost_base: Money | null;
  /** False is not a refusal: explicit linking is the documented override. */
  within_tolerance: boolean;
}

export interface TransferCandidates {
  items: TransferCandidate[];
}

export interface TransferLink {
  from_txn_id: UUID;
  to_txn_id: UUID;
}

export interface Transfer {
  transfer_group_id: UUID;
  matched_by: string;
  fx_cost_base: Money | null;
  txn_ids: UUID[];
}

export interface TransferDetail {
  transfer_group_id: UUID;
  matched_by: string;
  fx_cost_base: Money | null;
  /** Ordered by date then id; pick the side you want by the amount's sign. */
  legs: Transaction[];
}

/** Candidate legs for one transaction. Disabled until a transaction is named. */
export function useTransferCandidates(txnId: string | null, days = 5) {
  return useQuery({
    queryKey: ["transfer-candidates", txnId, days],
    queryFn: () =>
      api.get<TransferCandidates>(
        `/transactions/transfer-candidates?txn_id=${encodeURIComponent(txnId!)}&days=${days}`,
      ),
    enabled: !!txnId,
  });
}

/**
 * A linked pair, read whole. A leg carries only its group id, so this is how the
 * detail sheet finds the *other* leg — and the residual the pair cost.
 */
export function useTransfer(groupId: string | null) {
  return useQuery({
    queryKey: ["transfer", groupId],
    queryFn: () => api.get<TransferDetail>(`/transactions/transfers/${groupId}`),
    enabled: !!groupId,
  });
}

/** Both a link and an unlink change which rows are candidates for everything. */
function useInvalidateTransfers() {
  const invalidate = useInvalidateLedger();
  const qc = useQueryClient();
  return () => {
    invalidate();
    qc.invalidateQueries({ queryKey: ["transfer"] });
    qc.invalidateQueries({ queryKey: ["transfer-candidates"] });
  };
}

export function useLinkTransfer() {
  const invalidate = useInvalidateTransfers();
  return useMutation({
    mutationFn: (body: TransferLink) => api.post<Transfer>("/transactions/transfers", body),
    onSuccess: invalidate,
  });
}

export function useUnlinkTransfer() {
  const invalidate = useInvalidateTransfers();
  return useMutation({
    mutationFn: (groupId: string) => api.del<void>(`/transactions/transfers/${groupId}`),
    onSuccess: invalidate,
  });
}
